### PR TITLE
Intern Vec<ProgramClause<I>>

### PR DESCRIPTION
**Pull request by tirr-c**
This PR interns `ProgramClause` and `Vec<ProgramClause>`, and makes changes accordingly.

### CODE REVIEW

File: chalk-ir/src/lib.rs [hunk 1]
```diff
@@ -1227,17 +1227,42 @@ impl<I: Interner> ProgramClauseImplication<I> {
...
+    pub fn intern(self, interner: &I) -> ProgramClause<I> {
+        ProgramClause {
+            clause: interner.intern_program_clause(self),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasInterner)]
+pub struct ProgramClause<I: Interner> {
+    clause: I::InternedProgramClause,
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs:1254]: Nit: I think we we call this `interned` in all the other structs

File: chalk-ir/src/lib.rs [hunk 2]
```diff
@@ -1227,17 +1227,42 @@ impl<I: Interner> ProgramClauseImplication<I> {
...
+}
+
+impl<I: Interner> ProgramClause<I> {
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs:1257]: Nit: We should probably have a `new` function, like [the `Goal::new`](http://rust-lang.github.io/chalk/chalk_ir/struct.Goal.html#method.new)

File: chalk-ir/src/lib.rs [hunk 3]
```diff
@@ -1227,17 +1227,42 @@ impl<I: Interner> ProgramClauseImplication<I> {
...
+    pub fn into_from_env_clause(self, interner: &I) -> ProgramClause<I> {
+        let program_clause_data = self.data(interner);
+        let new_clause = program_clause_data.clone().into_from_env_clause(interner);
+        ProgramClause {
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs:1261]: then we can call `ProgramClause::new(...)`

File: chalk-ir/src/lib.rs [hunk 4]
```diff
@@ -1265,6 +1265,56 @@ impl<I: Interner> ProgramClause<I> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasInterner)]
+pub struct ProgramClauses<I: Interner> {
+    clauses: I::InternedProgramClauses,
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs:1273]: I expect we shoul call this `interned` too

File: chalk-ir/src/lib.rs [hunk 5]
```diff
@@ -1227,17 +1227,42 @@ impl<I: Interner> ProgramClauseImplication<I> {
...
+        let new_clause = program_clause_data.clone().into_from_env_clause(interner);
+        ProgramClause {
+            clause: interner.intern_program_clause(new_clause),
+        }
+    }
+
+    pub fn data(&self, interner: &I) -> &ProgramClauseData<I> {
+        interner.program_clause_data(&self.clause)
+    }
 }
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs:1269]: Also I think we should add an `interned(&self) -> &I::InternedProgramClause` accessor

**Comment by nikomatsakis** [comment]
Looks like [the tests need to be updated, too](https://github.com/rust-lang/chalk/runs/559004396#step:6:290)

**Comment by tirr-c** [comment]
@nikomatsakis Done, pushed changes!

File: chalk-ir/src/lib.rs (hunk 1 above)
Comment (jackh726) [reply chalk-ir/src/lib.rs]: We call it `interned` except in `Goal` and `Goals` IIRC. I think those should be renamed and everything should be `interned`.

Comment (nikomatsakis) [reply chalk-ir/src/lib.rs]: It looks like @tirr-c fixed the new cases in this PR, we can create an issue for the others

**Pull request merged**